crates/net/network/src/transactions/mod.rs

```diff
@@ -1082,84 +1082,113 @@ where
-                        // known that this transaction is bad. (e.g. consensus
-                        // rules)
-                        if err.is_bad_transaction() && !this.network.is_syncing() {
-                            debug!(target: "net::tx", ?err, "bad pool transaction import");
-                            this.on_bad_import(err.hash);
-                            continue
+                        debug!(target: "net::tx", ?err, "bad pool transaction batch import");
+                        continue
+                    }
+                };
+
+                for res in import_res {
+                    match res {
+                        Ok(hash) => {
+                            this.on_good_import(hash);
+                        }
+                        Err(err) => {
+                            // if we're _currently_ syncing and the transaction is bad we ignore it,
+                            // otherwise we penalize the peer that sent the bad
+                            // transaction with the assumption that the peer should have
+                            // known that this transaction is bad. (e.g. consensus
+                            // rules)
+                            if err.is_bad_transaction() && !this.network.is_syncing() {
+                                debug!(target: "net::tx", ?err, "bad pool transaction import");
+                                this.on_bad_import(err.hash);
+                                continue
+                            }
+                            this.on_good_import(err.hash);
                         }
-                        this.on_good_import(err.hash);
                     }
                 }
+                some_ready = true;
             }
-        }
 
-        // handle and propagate new transactions
-        let mut new_txs = Vec::new();
-        while let Poll::Ready(Some(hash)) = this.pending_transactions.poll_next_unpin(cx) {
-            new_txs.push(hash);
-        }
-        if !new_txs.is_empty() {
-            this.on_new_transactions(new_txs);
-        }
+            // handle and propagate new transactions
+            let mut new_txs = Vec::new();
+            if let Poll::Ready(Some(hash)) = this.pending_transactions.poll_next_unpin(cx) {
+                new_txs.push(hash);
+                some_ready = true;
+            }
+            if !new_txs.is_empty() {
+                this.on_new_transactions(new_txs);
+            }
 
-        // all channels are fully drained and import futures pending
+            // all channels are fully drained and import futures pending
+            if !some_ready {
+                break
+            }
+
+            budget -= 1;
+            // If the budget is exhausted we manually yield back control to tokio. See
+            // `NetworkManager` for more context on the design pattern.
+            if budget == 0 {
+                break
```

Comment:
this should call cx.waker().wake_by_ref(); and return

Reply:
that is what happens, it breaks from the loop then on line 1191 schedules for wake up and returns pending

Reply:
it does, when the loop breaks. same thing happens there is no more work.

Reply:
this is only needed when we exit due to exhausted budget

Reply:
I believe it is also needed when we exit because there is no more work, since this future is spawned in a thread, and hence its `poll` function will not be manually called again

Reply:
ok, we let the tokio timer handle it, I follow. would be nice to control this timer...